python/ray/ray_process_reaper.py

```diff
@@ -0,0 +1,48 @@
+import os
+import signal
+import sys
+import time
+"""
+This is a lightweight "reaper" process used to ensure that ray processes are
+cleaned up properly when the main ray process dies unexpectedly (e.g.,
+segfaults or gets SIGKILLed). Note that processes may not be cleaned up
+properly if this process is SIGTERMed or SIGKILLed.
+
+It detects that its parent has died by reading from stdin, which must be
+inherited from the parent process so that the OS will deliver an EOF if the
+parent dies. When this happens, the reaper process kills the rest of its
+process group (first attempting graceful shutdown with SIGTERM, then escalating
+to SIGKILL).
+"""
+
+SIGTERM_GRACE_PERIOD_SECONDS = 1
+
+
+def reap_process_group(*args):
+    def sigterm_handler(*args):
+        # Give a one-second grace period for other processes to clean up.
+        time.sleep(SIGTERM_GRACE_PERIOD_SECONDS)
+        # SIGKILL the pgroup (including ourselves) as a last-resort.
+        os.killpg(0, signal.SIGKILL)
+
+    # Set a SIGTERM handler to handle SIGTERMing ourselves with the group.
+    signal.signal(signal.SIGTERM, sigterm_handler)
+
+    # Our parent must have died, SIGTERM the group (including ourselves).
+    os.killpg(0, signal.SIGTERM)
```

Comment:
Does Ray support windows? `os.killpg` is unix only. An alternate route would be to avoid process groups, and pass the process reaper the PIDs to be killed, which should work cross platform.

Reply:
We're currently unix-only. I considered that option, but it leaves a race condition if you either spawn the process reaper last or first and pass it PIDs of other processes - the main process could be killed between spawning other processes and spawning/notifying the reaper.

Reply:
Since it's needed for the release, I think it's fine to merge without cross platform support. However, @mehrdadn is working to support Windows right now, so we will want to change the behavior in the near future.

Reply:
@robertnishihara thanks for the heads up! I can look into Windows support when I get to this, but in the meantime @edoakes when you come across POSIX-only situations like this, can you check if `sys.platform == 'win32'` and if so maybe raise something like `NotImplementedError("Windows not yet supported")` (even maybe with `TODO(mehrdadn)`)? It'll help me avoid a lot of headaches if things fail loudly.

Reply:
Added a TODO, the `NotImplementedError` wouldn't be very useful in this case because it isn't surfaced anywhere for this process.

Reply:
Okay thanks.